### PR TITLE
[PARTI-542] Autocomplete UI Library: Handle permissions on protected branches and implement `npm version` already existing logic

### PR DESCRIPTION
**Pull request by guiquintelas**
## Changes
- Changes both `package.json` and `package-lock.json` version fields to `0.0.0`.
- Update both `publish.yml` and `publish-npm.yml` workflows to track the library's version using git tags and releases.
- Change the `git push` to only push the latest created tag.
- Fix final Slack notification wrong status.
- Adds a `Deploy Storybook` workflow.
- Adds a `Publish to CDN` workflow.
- Changes `publish.yml` to:
  - Also deploy all storybook changes to GH pages.
  - Updates the `version.ts` file before publishing the new NPM version.

Some [context](https://constructor.slack.com/archives/C04B29BNW9H/p1699374608096719) regarding the `0.0.0` version change.

Feel free to test these workflow changes using this [fork](https://github.com/Constructor-io/parti-505-constructorio-ui-autocomplete/actions/workflows/publish.yml).

### CODE REVIEW

File: README.md
Comment (pedro-lb) [inline README.md:217]: **Suggestion:** The repo is public, so let's remove the slack channel name

```suggestion
6. Report the progress on the [relevant Slack channel](https://constructor.slack.com/archives/C061D3CFVR9).
```

File: .github/workflows/publish.yml
Comment (pedro-lb) [inline .github/workflows/publish.yml:61]: **Typo**

```suggestion
            # remove the "v" prefix.
```

File: .github/workflows/helpers/map-strategy.js
Comment (pedro-lb) [inline .github/workflows/helpers/map-strategy.js:19]: **Suggestion:** Instead of having this file, I think we should simply have the same values in the dropdown - so that we don't need to map anything in a js script.

Think that this workflow will be reused in other places, so the simpler it is the better it is to reuse 🥳

**Review by pedro-lb — CHANGES_REQUESTED** [review]
Awesome work here @guiquintelas! 🔥🔥 

Left some reviews, I think we'll be good to go once those are solved!

File: README.md
Comment (guiquintelas) [reply README.md:217]: That's right! Nice catch!

File: .github/workflows/publish-npm.yml
Comment (pedro-lb) [inline .github/workflows/publish-npm.yml:28]: **Issue:** we can drop this.

File: .github/workflows/publish-npm.yml
Comment (pedro-lb) [inline .github/workflows/publish-npm.yml:31]: **Suggestion:** We can drop this if we're receiving the version as an input.

Let's receive the version in the input and use it here

**Review by pedro-lb — CHANGES_REQUESTED** [review]

File: .github/workflows/publish-npm.yml
Comment (pedro-lb) [reply .github/workflows/publish-npm.yml:31]: Check this pr for an implementation suggestion:

- https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/104/files

File: .github/workflows/helpers/map-strategy.js
Comment (pedro-lb) [reply .github/workflows/helpers/map-strategy.js:19]: Note: we can use another action that maps 1:1 as discussed

Comment (guiquintelas) [reply .github/workflows/helpers/map-strategy.js]: Done! But sorry about the huge commit. I was continuously testing on a fork branch, so I just moved over all the changes at once 😬

File: .github/workflows/publish-npm.yml
Comment (guiquintelas) [reply .github/workflows/publish-npm.yml]: Done! ✅

File: .github/workflows/publish.yml
Comment (pedro-lb) [inline .github/workflows/publish.yml:28]: **Suggestion:** Should we deploy the storybook after publishing to npm and cdn? 

That way if something fails we'd avoid updating it

File: package.json
Comment (pedro-lb) [inline package.json:19]: **Question:** should we remove this?

**Review by pedro-lb — COMMENTED** [review]
Awesome work here @guiquintelas! 🔥🔥🔥🔥

Left a few comments, we should be good to go after solving those! 💯

File: .github/actions/create-version-tag/action.yml
Comment (guiquintelas) [reply .github/actions/create-version-tag/action.yml:49]: I thought about that. I tried to minimize possible developer errors that might break this workflow. In this case, the developer would forget about the new workflow and locally run a npm version command. This would guarantee that any wrongly created tag won't break this workflow.

Another upside I see with this approach is the ease of use to just re-run the main `Publish` workflow, in case the first step fails, instead of reaching into manual steps for everything else. The tag version is always based on the last release made, so if we're trying to create the same tag again, it's definitely because something went wrong, and we're trying to fix it.

But I might be missing some other edge cases that we might not want to use the `--force` flag. Wdyt?

Comment (pedro-lb) [reply .github/actions/create-version-tag/action.yml:49]: That makes sense! I think we're fine to use it. Let's leave it as is ✅

**Review by pedro-lb — APPROVED** [review]
LGTM! Awesome, awesome work here @guiquintelas! 🔥🔥🔥🔥

I think we're good to go!

File: .github/workflows/deploy-storybook.yml
Comment (esezen) [inline .github/workflows/deploy-storybook.yml:53]: Would you be able to explain how this works? Our docs (GH Pages) are served from `./docs` on main. How does this action deploy docs without having write access to main?

Comment (guiquintelas) [reply .github/workflows/deploy-storybook.yml:53]: For sure! 



GH provides two ways to _deploy_ using GH pages:
1. **From a branch**: Directly consumes all files right from the branch content - given the correct path.
2. **From GH Actions**: We're required to manually upload our files to GH.



We can perform this upload using the `actions/upload-pages-artifact` action. As described in this [GitHub article](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-with-a-custom-github-actions-workflow).

You can also see a working example over on [this repo](https://github.com/Constructor-io/connect-template-toolkit/blob/main/.github/workflows/deploy-docs-to-gh-pages.yml).

Finally, answering your question. We simply don't need to change any files on the `main` branch, GH stores all resulting built files as artifacts directly consumed by GH Pages.

File: .github/actions/create-version-tag/action.yml
Comment (esezen) [reply .github/actions/create-version-tag/action.yml:49]: Thanks for the explanation @guiquintelas !

> In this case, the developer would forget about the new workflow and locally run a npm version command.

They would have to push the created tag to main as well, right? And wouldn't the job override the existing tag in this case? In a way, it would be failing silently. There was something wrong with the tag, but the job fixed it without producing an error.

Would it make more sense to fail in this case? First of all, no one should push a tag to main. If they do, they need to remove it when they are aware (either on GH or locally via git). I think our git tags should always be "clean" and when something is wrong like a developer pushing a tag accidentally and running the job afterwards, I think the job should fail, explicitly call out the error, and someone needs to go to tags and clean it up before running the job again. 

-----------

Another scenario is, if I run the job with an existing version, would that mean it will override that tag? Again I think it might be safer to error out in this case. If someone is trying to push a version that already exists on NPM or in our git tags, that shouldn't be allowed or we would be re-writing history. 

(Let me know if I'm completely misunderstanding this)

File: .github/actions/create-version-tag/action.yml
Comment (esezen) [inline .github/actions/create-version-tag/action.yml:50]: This will only push the tag and nothing else, right?

File: .github/workflows/create-github-release.yml
Comment (esezen) [inline .github/workflows/create-github-release.yml:38]: Just out of curiosity, why are we not using the version number here instead of the commit?

File: .github/workflows/publish.yml
Comment (esezen) [inline .github/workflows/publish.yml:87]: Would the workflow run this step if `publish_to_npm` fails?

File: .github/workflows/publish-cdn.yml
Comment (esezen) [inline .github/workflows/publish-cdn.yml:53]: Does the first argument need `.js` extension? How does it work without the extension 🤔

**Review by esezen — COMMENTED** [review]
This is amazing @guiquintelas ! I love all the `jq` wizardry 🧙 
I left some comments, they are mostly me trying to understand things. Thanks for working on this, super excited to get this merged after BF 🚀🚀🚀

File: .github/actions/compile/action.yml
Comment (pedro-lb) [inline .github/actions/compile/action.yml:14]: **Typo**

```suggestion
    - name: Generate new src/version.ts file
```

File: .github/workflows/publish-cdn.yml
Comment (pedro-lb) [reply .github/workflows/publish-cdn.yml]: It does! My bad - forgot to add it.

Fixed in 10e4f3e6fb7a1555e27ea7869418f18547be0951.

File: .github/actions/compile/action.yml
Comment (pedro-lb) [reply .github/actions/compile/action.yml]: Fixed during the merge with my pr: f1dc3a43a4d91a86f57ff52ecf5d4fae90d447a0

File: .github/actions/compile/action.yml
Comment (pedro-lb) [reply .github/actions/compile/action.yml]: Fixed during the merge with my pr: f1dc3a43a4d91a86f57ff52ecf5d4fae90d447a0

File: .github/workflows/publish.yml
Comment (pedro-lb) [inline .github/workflows/publish.yml:86]: **Block:** Need to re-enable it once the CDN tests are done

File: .github/workflows/publish-cdn.yml
Comment (pedro-lb) [inline .github/workflows/publish-cdn.yml:57]: **Block:** make sure to get the correct s3 path + extension

**Review by pedro-lb — CHANGES_REQUESTED** [review]
Some things to remember on the cdn front

File: .github/workflows/publish.yml
Comment (pedro-lb) [reply .github/workflows/publish.yml]: 43fc868fed28aa04d90d87d74d0c706df8b28c31

File: .github/workflows/publish-cdn.yml
Comment (pedro-lb) [reply .github/workflows/publish-cdn.yml]: It's correct now

**Comment by pedro-lb** [comment]
Publish to CDN is working as expected:

image

image

File: .github/workflows/publish-cdn.yml
Comment (pedro-lb) [inline .github/workflows/publish-cdn.yml:58]: **Suggestion**: @guiquintelas small suggestion here, but maybe we could have an action (e.g. just like the `compile` one) to build this and throw it into the `env`?

I can see that we're doing the same thing in 3 places already, so it might be a bit flaky to maintain

**Review by pedro-lb — COMMENTED** [review]
Some more comments - we should be good to go after solving the pending ones!

File: README.md
Comment (guiquintelas) [reply README.md]: Nice catch! Typical copilot mistake 🤦 

194cee5f2d9f8c0b07a1e48245e9231ea4129291 Done ✅

File: .github/actions/compile/action.yml
Comment (guiquintelas) [reply .github/actions/compile/action.yml]: A 100% safe way, I'm not so sure. However, I made a test commit that performed a `cat` command after compiling the library. So we could see the resulting `.js` file created.

image

[Here is the full workflow execution](https://github.com/Constructor-io/parti-505-constructorio-ui-autocomplete/actions/runs/6936259743/job/18868096118).

File: .github/actions/create-version-tag/action.yml
Comment (guiquintelas) [reply .github/actions/create-version-tag/action.yml:50]: Yupp, that's right! 

[Source](https://stackoverflow.com/a/5195913/10745645)

File: .github/workflows/create-github-release.yml
Comment (guiquintelas) [reply .github/workflows/create-github-release.yml:38]: A new security best practice that I found out on the platform infrastructure's _GitHub Actions Security Cheat Sheet_ article. 

> When using third-party actions, pin the version with a commit hash rather than a tag to shield your workflow from potential supply-chain compromise:

> Tags can be moved or deleted, repeating SHA requires SHA-1 collision for valid Git object payload.

File: .github/workflows/publish.yml
Comment (guiquintelas) [reply .github/workflows/publish.yml:86]: It would. The reason was to streamline the happy path, making it as fast as possible. But the sad path would increase its troubleshooting complexity. 

I'm not sure if the tradeoffs are worth it. We would have a longer publishing process, but I'm pretty sure that's not so time-sensitive. But I know that any surprise complexity trying to debug a workflow error will only make the sad path sadder.

Done b6f94df5ad01006b6570d6e71ef639658aa93b6d ✅ 

@pedro-lb any concerns about running all actions sequentially?

File: .github/actions/create-version-tag/action.yml
Comment (guiquintelas) [reply .github/actions/create-version-tag/action.yml]: Totally agree @esezen.

I had the wrong mindset that the developer would just re-run the whole pipeline when encountering an error. That could cause many cascading errors. The best way is to run each subworkflow individually. And the best way to ensure that is to prevent the developer from re-running the main pipeline (supposing the error happened after the tag creation - most likely).

I imagine the two most probable actions to fail are:
1. Compiling the library.
2. Compiling the storybook.

The course of action would then be fixing the error and:
1. Manually re-running:
  a. Publish to NPM.
  b. Publish to CDN.
  c. Deploy storybook (if needed).
2. Just re-run the "Deploy storybook" action.

In both cases, we wouldn't be able to re-run the main pipeline because the GH release would have already been created. 

> I think the job should fail, explicitly call out the error, and someone needs to go to tags and clean it up before running the job again.

This 👌 

Thanks for shining a light on this @esezen!

File: .github/workflows/publish.yml
Comment (pedro-lb) [reply .github/workflows/publish.yml:86]: No concerns at all. I think it's easier to manage if we run it sequentially.

File: .github/workflows/publish-cdn.yml
Comment (guiquintelas) [reply .github/workflows/publish-cdn.yml]: I tried implementing with a shared `env`, but we would lose our failsafe when manually running jobs. If anyone accidentally fills in the version with a prefix, we can safely handle it.

So I went ahead and created a new action just for it.

Done 6ce74b6017d7ff9b47ea0256031a4d69cffc8bae ✅

File: .github/actions/remove-version-prefix/action.yml
Comment (pedro-lb) [inline .github/actions/remove-version-prefix/action.yml:21]: **Issue:** I think this comment is wrong.

Maybe we can update it to something like `Adds the version to output`

**Review by pedro-lb — APPROVED** [review]
LGTM! I left a minor discussion but I think we're ready to go after solving it.

Awesome work here, @guiquintelas! 🚀 🔥

File: .github/actions/remove-version-prefix/action.yml
Comment (guiquintelas) [reply .github/actions/remove-version-prefix/action.yml]: Oops, done c13cd57dcd8c858cb3007278eba6058edaf06394

**Review by pedro-lb — APPROVED** [review]
LGTM!

**Pull request merged**
